### PR TITLE
Audio2Text: model init, test on golos ds

### PR DESCRIPTION
**Pull request by GlebNeshchetkin**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -4,4 +4,4 @@ pillow==11.1.0
 typed-argument-parser==1.10.1
 transformers==4.49.0
 torch==2.6.0
-accelerate==1.4.0
+accelerate==1.4.0
```
Comment (alexgiving) [inline requirements.txt]:
```suggestion
accelerate==1.4.0
tqdm==4.67.1
tqdm-stubs==0.2.1
```

File: src/audio2text/utils.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    """Generates a transcription for a single audio file."""
+    if os.path.exists(path_to_file):
+        speech_array, _ = librosa.load(path_to_file, sr=16_000)
+        inputs = processor(speech_array, sampling_rate=16_000, return_tensors="pt", padding=True)
+        with torch.no_grad():
+            logits = model(inputs.input_values, attention_mask=inputs.attention_mask).logits
+        predicted_ids = torch.argmax(logits, dim=-1)
+        transcription = processor.batch_decode(predicted_ids)[0]
+        return transcription
+    return "FileNotFoundError"
```
Comment (alexgiving) [inline src/audio2text/utils.py]: raise FileNotFoundError(f"No file {path_to_file}")

File: pyproject.toml [hunk 3]
```diff
@@ -28,7 +28,13 @@ disallow_untyped_defs = true
...
 module = [
     "transformers.*",
-    "accelerate.*"
+    "accelerate.*",
+    "librosa.*",
+    "soundfile.*",
+    "datasets.*",
+    "jiwer.*",
+    "scipy.*",
+    "typing.*"
```
Comment (alexgiving) [inline pyproject.toml]: Typing is the standard Python library. Remove it from ignore.

File: src/audio2text/utils.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
+import os
+import time
+from typing import cast
+
+import librosa  # pylint: disable=import-error
```
Comment (o-frolova) [inline src/audio2text/utils.py:5]: If you added this into .toml file. Could we remove "# pylint: disable=import-error"?

File: src/audio2text/utils.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+from transformers import Wav2Vec2ForCTC, Wav2Vec2Processor, WhisperForConditionalGeneration, WhisperProcessor
+
+
+def load_data_golos(path_to_audio: str) -> list[str]:
+    """Loads test dataset from GOLOS and saves audio samples to files."""
+    ds = load_dataset("bond005/sberdevices_golos_10h_crowd")
+    data_ = ds["test"]
+    real_transcriptions = []
+    for i, smpl in tqdm.tqdm(enumerate(data_)):
+        filename = path_to_audio + f"/output_{i}.wav"
```
Comment (o-frolova) [inline src/audio2text/utils.py]:
```suggestion
       filename = os.path.join(path_to_audio, f"output_{i}.wav")
```

File: src/audio2text/utils.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        # Audio(smpl["audio"]["array"], rate=smpl["audio"]["sampling_rate"])
+    return real_transcriptions
+
+
+def gen_transcriptions_golos(
+    model: Wav2Vec2ForCTC, processor: Wav2Vec2Processor, num_samples: int = 10, whisper_flag: bool = False
+) -> list[str]:
+    """Generates transcriptions for stored audio files using a speech model."""
+    transcriptions = []
+    audio_paths = [f"./audio_files/output_{i}.wav" for i in range(num_samples)]
```
Comment (o-frolova) [inline src/audio2text/utils.py]: Can we pass the folder path as a function argument? And then use os.path.join to this line

File: src/audio2text/utils.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    if os.path.exists(path_to_file):
+        speech_array, _ = librosa.load(path_to_file, sr=16_000)
+        inputs = processor(speech_array, sampling_rate=16_000, return_tensors="pt", padding=True)
+        with torch.no_grad():
+            logits = model(inputs.input_values, attention_mask=inputs.attention_mask).logits
+        predicted_ids = torch.argmax(logits, dim=-1)
+        decoded = processor.batch_decode(predicted_ids)
+        transcription = cast(str, decoded[0]) if decoded else ""
+        return transcription
+    raise FileNotFoundError(f"No file {path_to_file}")
```
Comment (o-frolova) [inline src/audio2text/utils.py]: Can we rewrite this using try-except?

File: src/audio2text/utils.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    """Generates a transcription for a single audio file."""
+    if os.path.exists(path_to_file):
+        audio_input, _ = librosa.load(path_to_file, sr=16000)
+        input_features = processor(audio_input, return_tensors="pt").input_features
+        with torch.no_grad():
+            generated_ids = model.generate(input_features)
+        decoded = processor.batch_decode(generated_ids[0], skip_special_tokens=True)
+        transcription = cast(str, decoded) if decoded else ""
+        return transcription
+    raise FileNotFoundError(f"No file {path_to_file}")
```
Comment (o-frolova) [inline src/audio2text/utils.py]: Can we rewrite this using try-except?

File: src/audio2text/utils.py (hunk 4 above)
Comment (GlebNeshchetkin) [reply src/audio2text/utils.py:5]: @alexgiving

File: config/run_mypy.sh [hunk 9]
```diff
@@ -2,5 +2,7 @@
 
 source config/bootstrap.sh
 
+mypy --install-types
+
 mypy config/
-mypy src/
+mypy src/
```
Comment (alexgiving) [inline config/run_mypy.sh]: Add new line, please

File: src/audio2text/audio_files/output_321.wav
Comment (alexgiving) [inline src/audio2text/audio_files/output_321.wav:1]: Rm file

File: src/audio2text/example.py [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+from src.audio2text.utils import (
+    calc_test_metrics,
+    gen_transcription,
+    gen_transcription_whisper,
+    gen_transcriptions_golos,
+    load_data_golos,
+)
+
+# load whisper model and processor
```
Comment (alexgiving) [inline src/audio2text/example.py]: Use main please

```suggestion
if __name__ == "__main__":
```

File: src/audio2text/utils.py (hunk 4 above)
Comment (alexgiving) [reply src/audio2text/utils.py:5]: Agree. Ignore should be removed here

File: src/audio2text/utils.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    whisper_flag: bool = False,
+) -> list[str]:
+    """Generates transcriptions for stored audio files using a speech model."""
+    transcriptions = []
+    audio_paths = [os.path.join(folder_to_audio, f"output_{i}.wav") for i in range(num_samples)]
+    for audio_path in audio_paths:
+        if whisper_flag:
+            transcription = gen_transcription_whisper(model, processor, audio_path)
+        else:
+            transcription = gen_transcription(model, processor, audio_path)
```
Comment (alexgiving) [inline src/audio2text/utils.py:42]: Todo: make a factory of transcriptions

File: src/audio2text/utils.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    print("WER:", wer(real_transcriptions[:len_], generated_transcriptions[:len_]))
+    print("CER:", cer(real_transcriptions[:len_], generated_transcriptions[:len_]))
+
+
+def gen_transcription(model: Wav2Vec2ForCTC, processor: Wav2Vec2Processor, path_to_file: str) -> str:
+    """Generates a transcription for a single audio file."""
+    try:
+        speech_array, _ = librosa.load(path_to_file, sr=16_000)
+        inputs = processor(speech_array, sampling_rate=16_000, return_tensors="pt", padding=True)
+        with torch.no_grad():
```
Comment (alexgiving) [inline src/audio2text/utils.py]: Is it really used? You are not in training mode

File: src/audio2text/utils.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+            logits = model(inputs.input_values, attention_mask=inputs.attention_mask).logits
+        predicted_ids = torch.argmax(logits, dim=-1)
+        decoded = processor.batch_decode(predicted_ids)
+        transcription = cast(str, decoded[0]) if decoded else ""
+        return transcription
+    except FileNotFoundError as exc:
+        raise FileNotFoundError(f"No file {path_to_file}") from exc
```
Comment (alexgiving) [inline src/audio2text/utils.py]: File validation should be performed before try except

File: src/audio2text/utils.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    except Exception as e:
+        raise RuntimeError(f"Error during transcription: {e}") from e
```
Comment (alexgiving) [inline src/audio2text/utils.py]: Looks like this try except construction could be removed without influence on code

File: src/audio2text/utils.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+
+
+def gen_transcription_whisper(
+    model: WhisperForConditionalGeneration, processor: WhisperProcessor, path_to_file: str
+) -> str:
+    """Generates a transcription for a single audio file."""
+    try:
```
Comment (alexgiving) [inline src/audio2text/utils.py]: Same concerns

**Review by alexgiving — CHANGES_REQUESTED** [review]
In general, remove all commented lines

**Review by alexgiving — DISMISSED** [review]

**Review by o-frolova — APPROVED** [review]

**Review by alexgiving — APPROVED** [review]
